File: org.opencompare/api-java-impl/src/main/java/org/opencompare/api/java/impl/io/KMFJSONLoader.java
package org.opencompare.api.java.impl.io;

import org.opencompare.api.java.*;
import org.opencompare.api.java.impl.PCMImpl;
import org.opencompare.api.java.io.PCMLoader;
import org.kevoree.modeling.api.KMFContainer;
import org.kevoree.modeling.api.json.JSONModelLoader;
import pcm.factory.DefaultPcmFactory;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by gbecan on 12/12/14.
 */
public class KMFJSONLoader implements PCMLoader {

    private DefaultPcmFactory kpcmFactory = new DefaultPcmFactory();
    private JSONModelLoader loader = kpcmFactory.createJSONLoader();
    private PCMBase64Encoder encoder = new PCMBase64Encoder();

    @Override
    public List<PCMContainer> load(String json) {
        List<KMFContainer> containers = loader.loadModelFromString(json);
        return load(containers);
    }

    @Override
    public List<PCMContainer> load(File file) throws IOException {
        byte[] bytes = Files.readAllBytes(file.toPath());
        String json = new String(bytes, StandardCharsets.UTF_8);
        return load(json);
    }

    private List<PCMContainer> load(List<KMFContainer> containers) {
        List<PCMContainer> containersPCM = new ArrayList<>();
        for (KMFContainer container : containers) {
            PCM pcm = new PCMImpl((pcm.PCM) container);
            encoder.decode(pcm);
            PCMContainer containerPCM = new PCMContainer();
            PCMMetadata metadata = new PCMMetadata(pcm);
            containerPCM.setPcm(pcm);
            containerPCM.setMetadata(metadata);
            containersPCM.add(containerPCM);
        }
        return containersPCM;
    }
}

File: org.opencompare/api-java-impl/src/main/java/org/opencompare/api/java/impl/io/KMFJSONLoader.scala
package org.opencompare.api.java.impl.io

import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.util

import org.opencompare.api.java.{PCMMetadata, PCMContainer}
import org.opencompare.api.java.impl.PCMImpl
import org.opencompare.api.java.io.PCMLoader
import org.opencompare.model.pcm.factory.DefaultPcmFactory

import collection.JavaConversions._

/**
 * Created by gbecan on 12/12/14.
 */
class KMFJSONLoader(val base64Decoding: Boolean = true) extends PCMLoader {

  private val kpcmFactory = new DefaultPcmFactory
  private val loader = kpcmFactory.createJSONLoader
  private val encoder = new PCMBase64Encoder

  def this() {
      this(true)
  }

  override def load(json: String): util.List[PCMContainer] = {
    val containers = loader.loadModelFromString(json).toList
    load(containers)
  }

  override def load(file: File): util.List[PCMContainer] = {
    val bytes = Files.readAllBytes(file.toPath())
    val json = new String(bytes, StandardCharsets.UTF_8)
    load(json)
  }

  private def load(containers : List[org.kevoree.modeling.api.KMFContainer] ) : List[PCMContainer] = {
    for (container <- containers) yield {
      val pcm = new PCMImpl(container.asInstanceOf[org.opencompare.model.PCM])
      encoder.decode(pcm)
      val containerPCM = new PCMContainer()
      val metadata = new PCMMetadata(pcm)
      containerPCM.setPcm(pcm)
      containerPCM.setMetadata(metadata)
      containerPCM
    }
  }
}

